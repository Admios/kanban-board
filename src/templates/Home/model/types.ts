export interface Todo {
  id: string;
  stateId: string;
  text: string;
  position: number;
  owner: string;
}

export interface State {
  id: string;
  name: string;
  position: number;
  color: string;
  owner: string;
}
