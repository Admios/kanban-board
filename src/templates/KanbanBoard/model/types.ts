export interface Task {
  id: string;
  stateId: string;
  text: string;
  position: number;
}

export interface State {
  id: string;
  boardId: string;
  name: string;
  position: number;
  color: string;
}

export interface Board {
  id: string;
  name: string;
  owner: string;
}
