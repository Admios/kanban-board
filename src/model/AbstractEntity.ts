/**
 * Entities represent how the data is stored in the database.
 */
export interface AbstractEntity extends Record<string, any> {}
