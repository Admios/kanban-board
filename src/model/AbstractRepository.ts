export interface AbstractRepository<T> {
  findById(id: string): Promise<T>;
  list(): Promise<T[]>;
  create(entity: T): Promise<T>;
  update(entity: T): Promise<T>;
  delete(entity: T): Promise<T>;
}
