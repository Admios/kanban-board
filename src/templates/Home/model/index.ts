import { StateDTO } from "@/model/State";
import { TaskDTO } from "@/model/Task";
import { UserDTO } from "@/model/User";
import { Immutable, produce } from "immer";
import { StateCreator, create } from "zustand";
import { State, Task } from "./types";

export type { State, Task } from "./types";

type HomeState = Immutable<{
  tasks: Record<string, Task[]>;
  states: Record<string, State>;
  user?: UserDTO;
}>;

interface HomeActions {
  initialize(
    initialTasks: TaskDTO[],
    initialStates: StateDTO[],
    initialUser?: UserDTO,
  ): void;
  addTask(newTask: Task): Task;
  moveTask(
    newTask: Task,
    fromStateId: string,
    toStateId: string,
    position: number,
  ): void;
  addState(newState: State): State;
  editTask(id: string, updatedValues: Partial<Task>): void;
  deleteTask: (id: string) => void;
}

const stateCreator: StateCreator<HomeState & HomeActions> = (set, get) => ({
  tasks: {},
  states: {},

  initialize(initialTasks, initialStates, initialUser) {
    const statesMap: Record<string, State> = {};
    const tasksMap: Record<string, Task[]> = {};

    initialStates.forEach((backendState) => {
      const result: State = {
        ...backendState,
        id: backendState.id,
        position: backendState.position,
      };
      statesMap[result.id] = result;
    });

    initialTasks.forEach((backendTask) => {
      const { stateId } = backendTask;
      if (!tasksMap[stateId]) {
        tasksMap[stateId] = [];
      }
      tasksMap[stateId][backendTask.position] = {
        ...backendTask,
        stateId,
        id: backendTask.id,
        position: backendTask.position,
      };
    });

    set({ tasks: tasksMap, states: statesMap, user: initialUser });
  },

  addTask: (newTask) => {
    const result = { ...newTask };
    const tasks = produce(get().tasks, (draft) => {
      if (!draft[newTask.stateId]) {
        draft[newTask.stateId] = [];
      }

      draft[newTask.stateId].push(result);
    });

    set({ tasks });
    return result;
  },

  moveTask: (newTask, fromStateId, toStateId, position) => {
    const tasks = produce(get().tasks, (draftState) => {
      const sourceState = (draftState[fromStateId] ?? []).filter(
        (task) => task.id !== newTask.id,
      );

      let destinationState = draftState[toStateId] ?? [];
      if (fromStateId === toStateId) {
        destinationState = sourceState;
      }

      if (position <= destinationState.length) {
        destinationState.splice(position, 0, newTask);
      } else {
        destinationState.push(newTask);
      }

      draftState[fromStateId] = sourceState;
      draftState[toStateId] = destinationState;
    });

    set({ tasks });
  },

  addState: (newState) => {
    const currentModel = get();
    const tasks = produce(currentModel.tasks, (draft) => {
      draft[newState.id] = [];
    });
    const states = produce(currentModel.states, (draft) => {
      const position = Object.keys(currentModel.states).length;
      draft[newState.id] = { ...newState, position };
    });

    set({ states, tasks });
    return states[newState.id];
  },

  editTask: (taskId, updatedValues) => {
    const currentState = get();
    const tasks = produce(currentState.tasks, (draft) => {
      for (const stateId in draft) {
        const stateTasks = draft[stateId];
        const taskIndex = stateTasks.findIndex((task) => task.id === taskId);
        if (taskIndex > -1) {
          stateTasks[taskIndex] = {
            ...stateTasks[taskIndex],
            ...updatedValues,
          };
          break;
        }
      }
    });

    set({ tasks });
  },

  deleteTask: (taskId) => {
    const currentState = get();
    const tasks = produce(currentState.tasks, (draft) => {
      for (const stateId in draft) {
        const stateTasks = draft[stateId];
        const taskIndex = stateTasks.findIndex((task) => task?.id === taskId);
        if (taskIndex > -1) {
          stateTasks.splice(taskIndex, 1);
          break;
        }
      }
    });

    set({ tasks });
  },
});

export const useZustand = create(stateCreator);
