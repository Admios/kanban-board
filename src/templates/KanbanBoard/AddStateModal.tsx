import { clsx } from "clsx";
import { KeyboardEventHandler, useState } from "react";
import { v4 as uuid } from "uuid";
import { addStateDB } from "./kanbanActions";
import { useZustand } from "./model";

interface AddModalProps {
  isOpen: boolean;
  onClose: () => void;
}

export function AddStateModal({ isOpen, onClose }: AddModalProps) {
  const states = useZustand((store) => store.states);
  const addState = useZustand((store) => store.addState);
  const user = useZustand((store) => store.user);
  const [title, setTitle] = useState("");
  const isError = title === "";

  const submit = () => {
    if (!user) {
      return;
    }
    const newState = {
      name: title,
      id: uuid(),
      color: "black",
      position: Object.values(states).length,
      owner: user.email,
    };
    addStateDB(newState);
    addState(newState);
    handleClose();
  };

  const handleClose = () => {
    onClose();
    setTitle("");
  };

  const submitOnEnter: KeyboardEventHandler<HTMLInputElement> = (event) => {
    if (event.key === "Enter" && !isError) {
      submit();
    }
  };

  return (
    <div
      className={clsx("modal", isOpen && "is-active")}
      role="dialog"
      aria-labelledby="add-state-modal-title"
    >
      <div className="modal-background" onClick={handleClose} />
      <div className="modal-card">
        <header className="modal-card-head">
          <p id="add-state-modal-title" className="modal-card-title">
            Add State
          </p>
          <button
            className="delete"
            aria-label="close"
            onClick={handleClose}
          ></button>
        </header>
        <section className="modal-card-body">
          <div className="field">
            <label className="label">Name:</label>
            <div className="control">
              <input
                className={clsx("input", isError && "is-danger")}
                type="text"
                value={title}
                onChange={(event) => setTitle(event.target.value)}
                autoFocus
                onKeyUp={submitOnEnter}
              />
            </div>
            {!isError ? (
              <p className="help">The name of the new state.</p>
            ) : (
              <p className="help is-danger">Name is required.</p>
            )}
          </div>
        </section>
        <footer className="modal-card-foot">
          <button className="button" onClick={handleClose}>
            Close
          </button>
          <button
            className="button is-success"
            onClick={submit}
            disabled={isError}
          >
            Add State
          </button>
        </footer>
      </div>
    </div>
  );
}
