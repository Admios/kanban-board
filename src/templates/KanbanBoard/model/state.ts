import { StateDTO } from "@/model/State";
import { TaskDTO } from "@/model/Task";
import { UserDTO } from "@/model/User";
import { Immutable } from "immer";
import { State, Task } from "./types";

export type KanbanState = Immutable<{
  /**
   * The order of the states in the board.
   */
  statesOrder: string[];

  /**
   * The order of the tasks in each state.
   */
  tasksOrder: Record<string, string[]>;

  /**
   * Key: taskId. Value: Task.
   */
  tasks: Record<string, Task>;

  /**
   * Key: stateId. Value: State.
   */
  states: Record<string, State>;

  /**
   * The user that is currently logged in
   */
  user?: UserDTO;
}>;

export interface KanbanActions {
  setUser(user?: UserDTO): void;
  initialize(initialTasks: TaskDTO[], initialStates: StateDTO[]): void;
  addTask(newTask: Task): Task;
  moveTask(
    newTask: Task,
    fromStateId: string,
    toStateId: string,
    position: number,
  ): void;
  addState(newState: State): State;
  editTask(id: string, updatedValues: Partial<Task>): void;
  deleteTask: (id: string) => void;

  clear(): void;
}
