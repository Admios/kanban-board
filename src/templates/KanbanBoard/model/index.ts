import { produce } from "immer";
import { StateCreator, create } from "zustand";
import { State, Task } from "./types";
import { KanbanState, KanbanActions } from "./state";

export type { State, Task } from "./types";

const stateCreator: StateCreator<KanbanState & KanbanActions> = (set, get) => ({
  statesOrder: [],
  tasksOrder: {},
  tasks: {},
  states: {},

  initialize(initialTasks, initialStates) {
    const statesOrderSparseArray: string[] = [];
    const tasksOrder: Record<string, string[]> = {};
    const states: Record<string, State> = {};
    const tasks: Record<string, Task> = {};

    initialTasks.forEach((backendTask) => {
      tasks[backendTask.id] = backendTask;
      if (!tasksOrder[backendTask.stateId]) {
        tasksOrder[backendTask.stateId] = [];
      }
      // Warning: saved as sparse array! Compression is done later
      tasksOrder[backendTask.stateId][backendTask.position] = backendTask.id;
    });

    initialStates.forEach((backendState) => {
      states[backendState.id] = backendState;
      // Warning: saved as sparse array! compressed later
      statesOrderSparseArray[backendState.position] = backendState.id;
      // Compress sparse array
      tasksOrder[backendState.id] = (tasksOrder[backendState.id] ?? []).filter(
        (id) => !!id,
      );
    });

    // Compress sparse array
    const statesOrder = statesOrderSparseArray.filter((id) => !!id);

    set({
      statesOrder,
      tasksOrder,
      tasks,
      states,
    });
  },

  setUser(user) {
    set({ user });
  },

  addTask: (newTask) => {
    const tasks = produce(get().tasks, (draft) => {
      draft[newTask.stateId] = newTask;
    });

    const tasksOrder = produce(get().tasksOrder, (draft) => {
      if (!draft[newTask.stateId]) {
        draft[newTask.stateId] = [];
      }
      draft[newTask.stateId].push(newTask.id);
    });

    set({ tasks, tasksOrder });
    return newTask;
  },

  moveTask: (newTask, fromStateId, toStateId, position) => {
    const newTasksOrder = produce(get().tasksOrder, (draftState) => {
      const sourceState = (draftState[fromStateId] ?? []).filter(
        (task) => task !== newTask.id,
      );

      let destinationState = draftState[toStateId] ?? [];
      if (fromStateId === toStateId) {
        destinationState = sourceState;
      }

      if (position <= destinationState.length) {
        destinationState.splice(position, 0, newTask.id);
      } else {
        destinationState.push(newTask.id);
      }

      draftState[fromStateId] = sourceState;
      draftState[toStateId] = destinationState;
    });

    set({ tasksOrder: newTasksOrder });
  },

  addState: (newState) => {
    const tasksOrder = produce(get().tasksOrder, (draft) => {
      draft[newState.id] = [];
    });

    const statesOrder = produce(get().statesOrder, (draft) => {
      draft.push(newState.id);
    });

    const states = produce(get().states, (draft) => {
      draft[newState.id] = newState;
    });

    set({ states, tasksOrder, statesOrder });
    return states[newState.id];
  },

  editTask: (taskId, updatedValues) => {
    const tasks = produce(get().tasks, (draft) => {
      draft[taskId] = { ...draft[taskId], ...updatedValues };
    });

    set({ tasks });
  },

  deleteTask: (taskId) => {
    const { stateId } = get().tasks[taskId];
    const tasksOrder = produce(get().tasksOrder, (draft) => {
      draft[stateId] = draft[stateId].filter((id) => id !== taskId);
    });

    const tasks = produce(get().tasks, (draft) => {
      delete draft[taskId];
    });

    set({ tasks, tasksOrder });
  },
});

export const useZustand = create(stateCreator);
