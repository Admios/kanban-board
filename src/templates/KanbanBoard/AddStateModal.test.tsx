import { act, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { AddStateModal } from "./AddStateModal";
import { useZustand } from "./model";

jest.mock("./kanbanActions.ts");

function setupDialog() {
  act(() => {
    useZustand.setState({
      states: {},
      tasks: {},
      user: { email: "test" },
    });
  });
  const onClose = jest.fn();
  render(<AddStateModal isOpen={true} onClose={onClose} />);
  return { onClose };
}

function tearDownDialog() {
  act(() => useZustand.getState().clear());
}

it("should add a state when pressed", async () => {
  const { onClose } = setupDialog();

  await userEvent.type(screen.getByRole("textbox"), "My new state");
  await waitFor(() => {
    expect(screen.getByLabelText("Close").getAttribute("disabled")).toBeNull();
  });
  await userEvent.click(screen.getByText("Add State", { selector: "button" }));

  expect(onClose).toHaveBeenCalled();
  const states = useZustand.getState().states;
  expect(Object.values(states)).toHaveLength(1);
  const stateId = Object.keys(states)[0];
  expect(states[stateId]).toMatchSnapshot(
    {
      id: expect.any(String),
    },
    "States Result",
  );

  tearDownDialog();
});

it("should focus on the input component and submit when 'Enter' is pressed", async () => {
  const { onClose } = setupDialog();
  const textbox = screen.getByRole("textbox");

  expect(textbox).toHaveFocus();
  await userEvent.type(textbox, "My new state 2{enter}");
  expect(onClose).toHaveBeenCalled();
  const states = useZustand.getState().states;
  expect(Object.values(states)).toHaveLength(1);
  const stateId = Object.keys(states)[0];
  expect(states[stateId]).toMatchSnapshot(
    {
      id: expect.any(String),
    },
    "States Result",
  );

  tearDownDialog();
});
