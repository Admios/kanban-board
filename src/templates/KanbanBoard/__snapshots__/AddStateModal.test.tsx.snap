// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add a state when pressed: States Result 1`] = `
{
  "boardId": "",
  "color": "black",
  "id": Any<String>,
  "name": "My new state",
  "owner": "test",
  "position": 0,
}
`;

exports[`should focus on the input component and submit when 'Enter' is pressed: States Result 1`] = `
{
  "boardId": "",
  "color": "black",
  "id": Any<String>,
  "name": "My new state 2",
  "owner": "test",
  "position": 0,
}
`;
